File: sdks/python/stackhouse/vectors.py
"""Vector database client for Stackhouse Python SDK."""

import httpx
from typing import Optional, Dict, Any, List


class VectorClient:
    """Vector/embedding operations."""

    def __init__(self, client: httpx.AsyncClient):
        self._client = client

    async def create_collection(self, name: str, dimension: int = 384, distance: str = "cosine") -> Dict:
        resp = await self._client.post("/v1/vectors/collections", json={
            "name": name,
            "dimension": dimension,
            "distance": distance,
        })
        resp.raise_for_status()
        return resp.json()

    async def upsert(self, collection: str, documents: List[Dict]) -> Dict:
        """Upsert documents with auto-embedding."""
        resp = await self._client.post(f"/v1/vectors/{collection}/upsert", json={
            "documents": documents,
        })
        resp.raise_for_status()
        return resp.json()

    async def search(self, collection: str, query: str, top_k: int = 10,
                     filter: Optional[Dict] = None, hybrid: bool = False) -> Dict:
        """Semantic search (optionally hybrid with full-text)."""
        body = {
            "query": query,
            "top_k": top_k,
            "hybrid": hybrid,
        }
        if filter:
            body["filter"] = filter
        resp = await self._client.post(f"/v1/vectors/{collection}/search", json=body)
        resp.raise_for_status()
        return resp.json()

    async def delete(self, collection: str, ids: List[str]) -> Dict:
        resp = await self._client.post(f"/v1/vectors/{collection}/delete", json={"ids": ids})
        resp.raise_for_status()
        return resp.json()

    async def get(self, collection: str, ids: List[str]) -> Dict:
        resp = await self._client.post(f"/v1/vectors/{collection}/get", json={"ids": ids})
        resp.raise_for_status()
        return resp.json()

    async def list_collections(self) -> Dict:
        resp = await self._client.get("/v1/vectors/collections")
        resp.raise_for_status()
        return resp.json()

File: js-sdks/stackhouse-js/src/lib/Vectors.ts
/**
 * VectorClient - Vector search operations for Stackhouse
 * Powered by pgvector for similarity search with HNSW indexing
 */

export interface VectorSearchOptions {
    /** Number of results to return (default: 10) */
    topK?: number;
    /** Distance metric: 'cosine' | 'l2' | 'inner_product' (default: 'cosine') */
    metric?: 'cosine' | 'l2' | 'inner_product';
    /** Optional metadata filters */
    filters?: Record<string, any>;
    /** Vector column name (default: 'embedding') */
    column?: string;
}

export interface VectorUpsertOptions {
    /** Optional ID for update (omit for insert) */
    id?: number;
    /** Additional data to store alongside the vector */
    data?: Record<string, any>;
    /** Vector column name (default: 'embedding') */
    column?: string;
}

export interface VectorSearchResult {
    id: number;
    similarity: number;
    data: Record<string, any>;
}

export interface VectorInfo {
    table: string;
    column: string;
    dimensions: number;
    index_type: string;
    row_count: number;
}

export class VectorClient {
    constructor(
        private url: string,
        private headers: Record<string, string>
    ) { }

    /**
     * Get a vector query builder for a specific collection
     */
    collection(name: string): VectorCollectionClient {
        return new VectorCollectionClient(this.url, name, this.headers);
    }
}

export class VectorCollectionClient {
    constructor(
        private url: string,
        private collectionName: string,
        private headers: Record<string, string>
    ) { }

    /**
     * Perform a similarity search
     * @param queryVector The query embedding vector
     * @param options Search options (topK, metric, filters)
     */
    async search(queryVector: number[], options?: VectorSearchOptions): Promise<{
        success: boolean;
        data: VectorSearchResult[];
        count: number;
    }> {
        const response = await fetch(
            `${this.url}/v1/vectors/${this.collectionName}/search`,
            {
                method: 'POST',
                headers: this.headers,
                body: JSON.stringify({
                    vector: queryVector,
                    top_k: options?.topK ?? 10,
                    metric: options?.metric ?? 'cosine',
                    filters: options?.filters,
                    column: options?.column ?? 'embedding',
                }),
            }
        );

        if (!response.ok) {
            const error = await response.json().catch(() => ({}));
            throw new Error(error?.error?.message || `Vector search failed: ${response.status}`);
        }

        return response.json();
    }

    /**
     * Insert or update a record with a vector embedding
     * @param embedding The embedding vector
     * @param options Upsert options (id, data, column)
     */
    async upsert(embedding: number[], options?: VectorUpsertOptions): Promise<{
        success: boolean;
        data: { id: number; collection: string; dimensions: number };
        message: string;
    }> {
        const response = await fetch(
            `${this.url}/v1/vectors/${this.collectionName}/upsert`,
            {
                method: 'POST',
                headers: this.headers,
                body: JSON.stringify({
                    embedding,
                    id: options?.id,
                    data: options?.data,
                    column: options?.column ?? 'embedding',
                }),
            }
        );

        if (!response.ok) {
            const error = await response.json().catch(() => ({}));
            throw new Error(error?.error?.message || `Vector upsert failed: ${response.status}`);
        }

        return response.json();
    }

    /**
     * Batch insert/update records with vector embeddings
     */
    async batchUpsert(records: Array<{ embedding: number[]; id?: number; data?: Record<string, any> }>): Promise<{
        success: boolean;
        data: { ids: number[]; collection: string; count: number };
        message: string;
    }> {
        const response = await fetch(
            `${this.url}/v1/vectors/${this.collectionName}/batch`,
            {
                method: 'POST',
                headers: this.headers,
                body: JSON.stringify({
                    records: records.map(r => ({
                        embedding: r.embedding,
                        id: r.id,
                        data: r.data,
                        column: 'embedding',
                    })),
                }),
            }
        );

        if (!response.ok) {
            const error = await response.json().catch(() => ({}));
            throw new Error(error?.error?.message || `Batch upsert failed: ${response.status}`);
        }

        return response.json();
    }

    /**
     * Get vector column metadata for this collection
     */
    async info(): Promise<{ success: boolean; data: VectorInfo[] }> {
        const response = await fetch(
            `${this.url}/v1/vectors/${this.collectionName}/info`,
            { headers: this.headers }
        );

        if (!response.ok) {
            const error = await response.json().catch(() => ({}));
            throw new Error(error?.error?.message || `Vector info failed: ${response.status}`);
        }

        return response.json();
    }
}
